pkg/certificate/configuration.go

```diff
@@ -2,169 +2,453 @@ package certificate
+}
+
+func caBundleName(webhookName, configName string) string {
+	return webhookName + "-" + configName
+}
+
+func namespacedHostname(name, namespace string) string {
+	return name + "." + namespace
+}
 
-	return clientConfigList[0].CABundle, nil
+func serviceHostname(name, namespace string) string {
+	return name + "." + namespace + serviceSubdomain
 }
 
-// getServicesFromConfiguration it retrieves all the references services at
-// webhook configuration clientConfig and in case there is URL instead of
-// ServiceRef it will reference fake one with webhook name, mgr namespace and
-// passing the url hostname at map value
-func (m *Manager) getServicesFromConfiguration(configuration client.Object) (map[types.NamespacedName][]string, error) {
+func serviceFqdn(name, namespace string) string {
+	return name + "." + namespace + serviceSubdomain + clusterDomain
+}
 
-	services := map[types.NamespacedName][]string{}
+func newCertificateIssue(name, namespace string) *chain.CertificateIssue {
+	commonName := serviceHostname(name, namespace)
+	hostnames := []string{
+		name,
+		namespacedHostname(name, namespace),
+		commonName,
+		serviceFqdn(name, namespace),
+	}
+	certificateBundle := chain.CertificateIssue{
+		Name:      commonName,
+		Hostnames: hostnames,
+		CACertPEM: make(map[string][]byte),
+	}
+	return &certificateBundle
+}
 
-	for _, clientConfig := range m.clientConfigList(configuration) {
+func mutatingWebhookConfig(webhook client.Object) *admissionregistrationv1.MutatingWebhookConfiguration {
+	return webhook.(*admissionregistrationv1.MutatingWebhookConfiguration)
+}
 
-		service := types.NamespacedName{}
-		hostnames := []string{}
+func validatingWebhookConfig(webhook client.Object) *admissionregistrationv1.ValidatingWebhookConfiguration {
+	return webhook.(*admissionregistrationv1.ValidatingWebhookConfiguration)
+}
 
-		if clientConfig.Service != nil {
-			service.Name = clientConfig.Service.Name
-			service.Namespace = clientConfig.Service.Namespace
-		} else if clientConfig.URL != nil {
-			service.Name = m.webhookName
-			service.Namespace = m.namespace
-			u, err := url.Parse(*clientConfig.URL)
-			if err != nil {
-				return nil, errors.Wrapf(err, "failed parsing webhook URL %s", *clientConfig.URL)
+// clientConfigMap returns the the list of webhooks's mutation or validating WebhookClientConfig
+//
+// The WebhookClientConfig type is share between mutating or validating so we can have a common function
+// that uses the interface runtime.Object and do some type checking to access it [1].
+//
+// [1] https://godoc.org/k8s.io/kubernetes/pkg/apis/admissionregistration#WebhookClientConfig
+func clientConfigMap(webhook client.Object) map[string]*admissionregistrationv1.WebhookClientConfig {
+	clientConfigMap := map[string]*admissionregistrationv1.WebhookClientConfig{}
+	switch webhook.(type) {
+	case *admissionregistrationv1.MutatingWebhookConfiguration:
+		mutatingWebhookConfig := mutatingWebhookConfig(webhook)
```

Comment:
do the casting directly here is the only place done and it's safe and now it was like this in the legacy code but there the "mutatingWebhookConfig(" function was call more than once.